pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,341 @@
+				d.networkForRegistry = n
+			}
+		}
+	}
+}
+
+func (d *MCPDiscovery) endpointNetwork(endpointIP string) string {
+	// If networkForRegistry is set then all endpoints discovered by this registry
+	// belong to the configured network so simply return it
+	if len(d.networkForRegistry) != 0 {
+		return d.networkForRegistry
+	}
+
+	// Try to determine the network by checking whether the endpoint IP belongs
+	// to any of the configure networks' CIDR ranges
+	if d.ranger == nil {
+		return ""
+	}
+	entries, err := d.ranger.ContainingNetworks(net.ParseIP(endpointIP))
+	if err != nil {
+		log.Errora(err)
+		return ""
+	}
+	if len(entries) == 0 {
+		return ""
+	}
+	if len(entries) > 1 {
+		log.Warnf("Found multiple networks CIDRs matching the endpoint IP: %s. Using the first match.", endpointIP)
+	}
+
+	return (entries[0].(namedRangerEntry)).name
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
+		notReadyIP, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		notReadyPort, err := strconv.Atoi(port)
+		if proxyIP != notReadyIP || err != nil {
+			continue
+		}
+		endpoints := make([]*model.IstioEndpoint, 0)
+		hostname := kube.ServiceHostname(conf.Name, conf.Namespace, d.options.DomainSuffix)
+		se, ok := conf.Spec.(*networking.ServiceEntry)
+		if !ok {
+			return nil, errors.New("getProxyServiceInstances: wrong type")
+		}
+
+		for _, h := range se.Hosts {
+			for _, svcPort := range se.Ports {
+				// add not ready endpoint
+				out = append(out, notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
+				// add other endpoints
+				out = append(out, serviceInstances(se, svcPort, conf, h)...)
+				var uid string
+				if d.options.mixerEnabled() {
+					uid = fmt.Sprintf("kubernetes://%s.%s", conf.Name, conf.Namespace)
+				}
+				// create an EDSEndpoint for notReady IP:Port
+				edsEndpoint := &model.IstioEndpoint{
+					Address:         notReadyIP,
+					EndpointPort:    uint32(notReadyPort),
+					ServicePortName: svcPort.Name,
+					Labels:          conf.Labels,
+					UID:             uid,
+					//	ServiceAccount:  ???, //TODO: nino-k
+					Network: d.endpointNetwork(notReadyIP),
+					//Locality:   ???, //TODO: nino-k
```

Comment:
This is in the ServiceEntry

Reply:
That locality is a little tricky,  I already get the locality for everything else from ServiceEntry endpoints  [e.g.](https://github.com/istio/istio/pull/12276/files#diff-eec332eccc20a758b8bb1e90490c3dc3R244) but that one is a for notReadyEndpoints, which means there is no iteration over the endpoints going on.

Reply:
Makes sense, thanks